### PR TITLE
store DateTime as nanoseconds in doc store

### LINKED ISSUE #2380: Store DateTime as nanos in docstore
we use `BinarySerializable` to encode datetimes into the doc store. Strangely, while DateTime are precise up to ns, we only store µs there. We should store the value we got with full precision instead 
https://github.com/quickwit-oss/tantivy/blob/99a59ad37e833e1129f7e1abd25738a0adfae73a/common/src/datetime.rs#L166-L176

### PR DESCRIPTION
**Pull request by PSeitz**
The doc store DateTime was truncated to microseconds previously. This
removes this truncation, while still keeping backwards compatibility.

This is done by passing the doc store version to the DocumentBinarySerializer and having different deserialization logic depending on the version for dates.

bump version format to 7.
add compat test to check the date time truncation.

closes https://github.com/quickwit-oss/tantivy/issues/2380

### CODE REVIEW

**Comment by fmassot** [comment]
@fulmicoton can you review it?

**Comment by fulmicoton** [comment]
Closes #2466 I assume

File: src/store/footer.rs [hunk 1]
```diff
@@ -26,15 +27,18 @@ impl BinarySerializable for DocStoreFooter {
 
     fn deserialize<R: io::Read>(reader: &mut R) -> io::Result<Self> {
         let doc_store_version = u32::deserialize(reader)?;
-        if doc_store_version != DOC_STORE_VERSION {
-            panic!("actual doc store version: {doc_store_version}, expected: {DOC_STORE_VERSION}");
+        if doc_store_version > DOC_STORE_VERSION {
```
Comment (fulmicoton) [inline src/store/footer.rs:30]: can we introduce a `DocStoreVersion` enum?

File: src/store/reader.rs [hunk 2]
```diff
@@ -203,8 +205,9 @@ impl StoreReader {
     pub fn get<D: DocumentDeserialize>(&self, doc_id: DocId) -> crate::Result<D> {
         let mut doc_bytes = self.get_document_bytes(doc_id)?;
 
-        let deserializer = BinaryDocumentDeserializer::from_reader(&mut doc_bytes)
-            .map_err(crate::TantivyError::from)?;
+        let deserializer =
+            BinaryDocumentDeserializer::from_reader(&mut doc_bytes, self.doc_store_version)
```
Comment (fulmicoton) [inline src/store/reader.rs:243]: yeah it is a very clean, future proof solution.

File: common/src/serialize.rs [hunk 3]
```diff
@@ -25,6 +25,63 @@ impl io::Write for Counter {
...
 
+/// The configuration for binary serialization.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct BinarySerializableConfig {
+    /// The precision with which DateTime is de/serialized
+    pub date_time_precision: DateTimePrecision,
+}
+
+/// Trait for configurable binary serialization
+pub trait ConfigurableBinarySerializable: fmt::Debug + Sized {
```
Comment (fulmicoton) [inline common/src/serialize.rs]: The introduction of  ConfigurableBinarySerializable seems quite overkill.

I understand this is an attempt to be future proof, but let's wait for the need to resurface 2 or 3 times before going there.

Also, I don't think common is the right place for this trait. We could have it closer to the BinaryDocumentSerializer

File: src/schema/document/se.rs [hunk 4]
```diff
@@ -94,56 +112,36 @@ where W: Write
...
                 }
                 ReferenceValueLeaf::Date(val) => {
-                    self.write_type_code(type_codes::DATE_CODE)?;
-                    val.serialize(self.writer)
-                }
-                ReferenceValueLeaf::Facet(val) => {
-                    self.write_type_code(type_codes::HIERARCHICAL_FACET_CODE)?;
-
-                    Cow::Borrowed(val).serialize(self.writer)
+                    self.serialize_with_type_code(type_codes::DATE_CODE, &val)
```
Comment (fulmicoton) [inline src/schema/document/se.rs]: I think we should just have some short date specific code here.

File: src/schema/document/de.rs [hunk 5]
```diff
@@ -416,57 +430,94 @@ where R: Read
...
+            &doc_store_version_to_serialize_config(self.doc_store_version)?,
+        )
+        .map(u64_to_f64)
+        .map_err(DeserializeError::from)
     }
 
     fn deserialize_datetime(self) -> Result<DateTime, DeserializeError> {
         self.validate_type(ValueType::DateTime)?;
-        <DateTime as BinarySerializable>::deserialize(self.reader).map_err(DeserializeError::from)
+        <DateTime as ConfigurableBinarySerializable>::deserialize(
```
Comment (fulmicoton) [inline src/schema/document/de.rs]: I think we should just ahve some short date specific code here.

**Review by fulmicoton — COMMENTED** [review]
Working on the DocumentBinarySerializer is a great idea. Introducing a ConfigurableBinarySerializer is overkill at this point however. Can we just push the logic on the date specific method in `de.rs`, `se.rs`.

File: src/store/footer.rs (hunk 1 above)
Comment (PSeitz) [reply src/store/footer.rs:30]: yep, I converted it to an enum

File: common/src/serialize.rs (hunk 3 above)
Comment (PSeitz) [reply common/src/serialize.rs]: Yes, it's kinda overkill currently, I removed it

File: src/schema/document/se.rs (hunk 4 above)
Comment (PSeitz) [reply src/schema/document/se.rs]: yes, I moved the version handling here

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**